File: 1986-largest-color-value-in-a-directed-graph/1986-largest-color-value-in-a-directed-graph.java
class Solution {
    /*
        topological sort (Kahn)
        copy colors value from parent node to child node
     */
    public int largestPathValue(String colors, int[][] edges) {
        int n = colors.length();
        int[] degrees = new int[n];
        HashMap<Integer, List<Integer>> map = new HashMap<>();
        for (int i = 0; i < n; i++)
            map.put(i, new ArrayList<>());
        for (int[] edge: edges) {
            int from = edge[0], to = edge[1];
            map.get(from).add(to);
            degrees[to]++;
        }
        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < n; i++) {
            if (degrees[i] == 0)
                queue.add(i);
        }
        int res = 0, nodeCount = 0;
        int[][] dp = new int[n][26];
        while (!queue.isEmpty()) {
            int p = queue.poll();
            nodeCount++;
            dp[p][colors.charAt(p) - 'a']++;    // max color count of all paths upto p
            res = Math.max(res, dp[p][colors.charAt(p) - 'a']);
            for (int next: map.get(p)) {
                for (int i = 0; i < 26; i++) {  // copy colors
                    dp[next][i] = Math.max(dp[next][i], dp[p][i]);
                }
                degrees[next]--;
                if (degrees[next] == 0)
                    queue.add(next);
            }
        }
        return nodeCount < n ? -1 : res;
    }
}

File: 1986-largest-color-value-in-a-directed-graph/1986-largest-color-value-in-a-directed-graph.rs
use std::collections::{HashMap, VecDeque};

impl Solution {
    pub fn largest_path_value(colors: String, edges: Vec<Vec<i32>>) -> i32 {
        let n = colors.len();
        let mut degrees = vec![0; n];
        let mut map: HashMap<i32, Vec<i32>> = HashMap::new();
        
        for i in 0..n {
            map.insert(i as i32, Vec::new());
        }
        
        for edge in &edges {
            let from = edge[0];
            let to = edge[1];
            map.get_mut(&from).unwrap().push(to);
            degrees[to as usize] += 1;
        }
        
        let mut queue: VecDeque<i32> = VecDeque::new();
        
        for i in 0..n {
            if degrees[i] == 0 {
                queue.push_back(i as i32);
            }
        }
        
        let mut res = 0;
        let mut node_count = 0;
        let mut dp = vec![vec![0; 26]; n];
        
        while !queue.is_empty() {
            if let Some(p) = queue.pop_front() {
                node_count += 1;
                dp[p as usize][colors.as_bytes()[p as usize] as usize - b'a' as usize] += 1;
                res = res.max(dp[p as usize][colors.as_bytes()[p as usize] as usize - b'a' as usize]);
                
                if let Some(neighbors) = map.get(&p) {
                    for &next in neighbors {
                        for i in 0..26 {
                            dp[next as usize][i] = dp[next as usize][i].max(dp[p as usize][i]);
                        }
                        degrees[next as usize] -= 1;
                        if degrees[next as usize] == 0 {
                            queue.push_back(next);
                        }
                    }
                }
            }
        }
        
        if node_count < n as i32 {
            -1
        } else {
            res
        }
    }
}